### PR TITLE
Factory pattern spike

### PR DESCRIPTION
**Pull request by thesheps**
@john-chapman - Introducing the factory pattern so as to minimise coupling.  A little refactoring here and there.  Again open for debate!

### CODE REVIEW

File: src/plang/ExpressionNode.h [hunk 1]
```diff
@@ -1,109 +1,24 @@
...
-		kTypeAdd,
-		kTypeSubtract,
-		kTypeCount
+		public:
+			enum Type
+			{
+				kTypeOperand = 0,
+				kTypeOperator
+			};
+
```
Comment (thesheps) [inline src/plang/ExpressionNode.h:18]: @john-chapman Ideally I'd like to see this enummy Type comparison stuff disappear, in favour of simple type comparison.  Couldn't figure out how to get that working, though!  All depends on whether or not we want to follow a strictly OO paradigm with the actual dev of the system..?

File: tests/plang/ExpressionNodeFactoryTests.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+			ExpressionNodeFactory factory;
+			ExpressionNode e1 = factory.create(tokstr.getNext());
+			ExpressionNode e2 = factory.create(tokstr.getNext());
+			ExpressionNode e3 = factory.create(tokstr.getNext());
+
+			THEN("Then the resultant ExpressionNodes are of the correct type") {
+				REQUIRE(e1.type == ExpressionNode::kTypeOperand);
+				REQUIRE(e2.type == ExpressionNode::kTypeOperator);
+				REQUIRE(e3.type == ExpressionNode::kTypeOperand);
+			}
```
Comment (thesheps) [inline tests/plang/ExpressionNodeFactoryTests.cpp:29]: @john-chapman - Just following on from the note I'd added to the ExpressionNode.h code file.  I'd love to see the type assertion here to be as simple as:

> `REQUIRE(e1 == OperandExpressionNode)`

or something...

**Comment by john-chapman** [comment]
@thesheps 

So what you're after is runtime type information. C++ has the typeid() operator, but it won't do what you want:

ExpressionNode operator = factory.create(tok);
if (typeid(e) == typeid(OperatorExpressionNode)) {
   std::cout << "It's an operator" << std::endl;
} else if (typeid(e) == typeid(OperandExpressionNode)) {
   std::cout << "It's an operand" << std::endl;
}

This will print nothing, because 'e' is neither an OperatorExpressionNode nor an OperandExpressionNode - it's just an ExpressionNode! This is because ExpressionNode is not dynamically polymorphic (it's type is bound at compiletime rather than runtime).

My opinion is that in this case using the language mechanisms for RTTI will be clunky compared to using simple enums, especially if we follow the route of having ExpressionNode being a POD (plain old data) type, which seems logical to me at the moment. So from that point of view, ExpressionNode is nothing more than a lump of data which we initialize from a token and feed in to the evaluation process. So somewhere in the evaluator will be a switch statement like this:

switch (node.getType()) {
case kOperator:
    DoSomethingWithOperatorNode(node);
    break;
case kOperand:
   DoSomethingwithOperandNode(node);
   break;
};

Which is fairly close to what you want with RTTI (and is almost certainly how dynamic dispatch is implemented under the hood anyway). For me, the stuff we've implemented so far is to achieve the goal of reflecting the data (the plang script), which I think favours a more straightforward 'flat' design.

**Comment by thesheps** [comment]
@john-chapman

Superb, that answers my challenge perfectly :) what are your thoughts as to
having the Operator and Operand derivatives of ExpressionNode nonetheless?
It would be good to enforce SRP wherever possible, but I wouldn't like the
type comparison overhead to be intrusive. A happy marriage of derived
expression node types and potentially an underlying enum type could be a
good call?

**Comment by john-chapman** [comment]
I would be inclined to say that using inheritance breaks the "plain old data" constraint on ExpressionNode without any real benefit. Also it prevents us using a union to store the actual value (which would be useful to keep the type small):

class ExpressionNode {
   NodeType m_type; // whether it's an operator or an operand
   union {
       int m_int;
       float m_float;
       OperandType m_operand; 
   } m_value;
};

Also I feel that in this case it would be nice to centralize the code which initializes an ExpressionNode from a token (and not spread it out over N different class constructors). I agree that this violates single responsibility _but_ I think that with small POD classes which have a known small number of subtypes, conflating those subtypes to be managed in one place is beneficial (and makes the code a little easier to reason about).

**Comment by thesheps** [comment]
That's fine - Happy to close this pull request with your challenge.  My
only concern is that ExpressionNode ends up knowing a lot about both
operands and operators.  If it starts to exhibit behaviours of multiple use
cases, I might re-create a new branch to this effect.  Sound ok? :)

**Comment by john-chapman** [comment]
Sounds ok! We just need to keep in mind the PODness of ExpressionNode - it should never be doing anything more than initializing its members from a token or returning queries about its type/value.

**Comment by thesheps** [comment]
Very good :)

**Comment by thesheps** [comment]
Closing this pull request as per discussion.

**Pull request closed without merging**